### PR TITLE
cifs: Mutichannel fixes for reconnect cases

### PR DESCRIPTION
**Pull request by sprasad-microsoft**
Yet to test out the changes. 
Raising a PR early for reviews.

### CODE REVIEW

File: fs/cifs/sess.c [hunk 1]
```diff
@@ -74,6 +74,21 @@ bool is_ses_using_iface(struct cifs_ses *ses, struct cifs_server_iface *iface)
 	return false;
 }
 
+unsigned long
+ses_get_chan_index(struct cifs_ses *ses,
+		   struct TCP_Server_Info *server)
+{
```
Comment (aaptel) [inline fs/cifs/sess.c:80]: if this function is exported used outside the file i would put a cifs_ prefix

File: fs/cifs/sess.c [hunk 2]
```diff
@@ -74,6 +74,21 @@ bool is_ses_using_iface(struct cifs_ses *ses, struct cifs_server_iface *iface)
...
+	int i;
+
+	for (i = 0; i < ses->chan_count; i++) {
+		if (ses->chans[i].server == server)
+			return i;
+	}
+
+	/* If we didn't find the channel, it is likely a bug */
```
Comment (aaptel) [inline fs/cifs/sess.c:88]: you can add a WARN_ON(1) here

File: fs/cifs/smb2pdu.c [hunk 3]
```diff
@@ -267,13 +267,13 @@ smb2_reconnect(__le16 smb2_command, struct cifs_tcon *tcon,
 	/*
 	 * If we are reconnecting an extra channel, bind
 	 */
-	if (server->is_channel) {
+	if (server->primary_server) {
 		ses->binding = true;
```
Comment (aaptel) [inline fs/cifs/smb2pdu.c:271]: I would add a helper inline function cifs_server_is_channel(server), makes it easier to read

File: fs/cifs/connect.c [hunk 4]
```diff
@@ -1334,6 +1356,10 @@ cifs_get_tcp_session(struct smb3_fs_context *ctx)
 		goto out_err_crypto_release;
 	}
 
+	if (primary_server) {
+		++primary_server->srv_count;
+		tcp_ses->primary_server = primary_server;
```
[the remark this answers is not in the document]
Comment (sprasad-microsoft) [reply fs/cifs/connect.c:1361]: I do this in cifs_put_tcp_session.

File: fs/cifs/sess.c (hunk 2 above)
Comment (sprasad-microsoft) [reply fs/cifs/sess.c:88]: Will do

File: fs/cifs/smb2pdu.c (hunk 3 above)
Comment (sprasad-microsoft) [reply fs/cifs/smb2pdu.c:271]: Will do.

**Pull request closed without merging**